sympy/sets/conditionset.py

```diff
@@ -33,6 +37,53 @@ class ConditionSet(Set):
     False
     >>> 5 in ConditionSet(x, x**2 > 4, S.Reals)
     True
+
+    If the value is not in the base set, the result is false:
+
+    >>> 5 in ConditionSet(x, x**2 > 4, Interval(2, 4))
+    False
+
+    Notes
+    =====
+
+    Symbols with assumptions should be avoided or else the
+    condition may evaluate without consideration of the set:
+
+    >>> n = Symbol('n', negative=True)
+    >>> cond = (n > 0); cond
+    False
+    >>> ConditionSet(n, cond, S.Integers)
+    EmptySet()
+
+    In addition, substitution of a dummy symbol can only be
+    done with a generic symbol with matching commutativity
+    or else a symbol that has identical assumptions.
+
+    Although expressions other than symbols may be used, this
+    is discouraged and will raise an error if the expression
+    is not found in the condition:
+
+    >>> ConditionSet(x + 1, x + 1 < 1, S.Integers)
```

Comment:
Expressions of this type may become confusing. `ConditionSet(x, x < 1, {-1, 0, 1})` is `{-1, 0}` but what should `ConditionSet(x + 1, x + 1 < 1, {-1, 0, 1})` be? I think that it could be expected to be "the set of `x + 1` for `x` in `{-1, 0, 1}` such that `x + 1 < 1`", which is `{0}`. To obtain `{-1, 0}`, the base set `{-1, 0, 1}` should be "relocated" to `{-2, -1, 0}`.

(This is analogous to replacing `x` by `x + 1` in `Integral(x, (x, -1, 1))`. The limits have to be changed too: `Integral(x + 1, (x, -2, 0))`.)

ConditionSets of this type may be useful in some connections, but I think that substitutions like `x -> x + 1` should be allowed only under strict restrictions (if at all).

Reply:
> I think that substitutions like x -> x + 1 should be allowed only under strict restrictions (if at all).

It is disallowed. While one can *create* a ConditionSet with a non-symbol (think "generator") a replacement cannot be made. 
```
>>> ConditionSet(x,x<1,S.Integers).subs(x,x+1)
ConditionSet(x, x < 1, S.Integers)
```
The elements in the base set are the values of the ConditionSet.sym that are filtered by the condition:
```
>>> ConditionSet(x+1,x+1<1,{0,1})
{0}
```
I don't really like that behavior. As much as I have done, however, is to make sure that the sym can be identified in the condition. When it can't, an error is raised as shown in the doctests. (But I see that the case of a FiniteSet base set is not handled properly to catch this.
```
>>> ConditionSet(x+1,x<2,{0,1,3})
ConditionSet(x + 1, x < 2, {0, 1, 3})  <==needs to raise an error.
```

Reply:
The definition is (from the docstring)

```
{x | condition(x) is True for x in S}
```

So the set `ConditionSet(x + 1, x + 1 < 1, {-1, 0, 1})` would be `{x + 1 | x + 1 < 1 and x + 1 in {-1, 0, 1}}`, which is the same as `{x | x < 1 and x in {-1, 0, 1}}` ({-1, 0}). 

Effectively, a condition set represents the set builder notation `{x | P(x)}`, but to disallow open ended definitions, a domain is required, `{x | P(x) ∧ x ∈ S }` (actually you could just let `S` be UniversalSet if you wanted).

Ideally we should allow more general set builder notation {f(x) | P}. But the problem is with the domain. Requiring it to be of the form `{f(x) | P ∧ f(x) ∈ S}` seems odd, and might be unexpected. 

I don't know how easy it would be to do, but can we make the third argument of the ConditionSet optional, defaulting to UniversalSet (and `Contains(x, UniversalSet)` should evaluate to True automatically)? Then allow more advanced set builder, but perhaps disallow it (I'm not sure) for the three argument variant. One can always include a domain restriction as part of the predicate, like `ConditionSet(x + 1, And(x + 1 < 0, Contains(x, {-1, 0, 1})))`.

I was never really clear on why the third argument to condition set was required. You can already create open ended sets anyway by using UniversalSet as the third argument. I think it would be more convenient, symbolically to just allow open ended sets. Forcing a domain makes it harder to deal with sets of more abstract things than just numbers. Obviously something like `solveset` should always return a set that is a subset of the given domain. 

By the way, we don't have symbolic sets yet, but once we do, it will be possible to build Russell's paradox with `ConditionSet(SetSymbol("A"), Not(Contains(SetSymbol("A"), SetSymbol("A")), S.UniversalSet)`.

Reply:
The [Wikipedia article](https://en.wikipedia.org/wiki/Set-builder_notation) is a good reference for this.

Reply:
How does this look?
```
>>> c = ConditionSet(x, x < 2); c
ConditionSet(x, x < 2)
>>> c.base_set
UniversalSet()
>>> pprint(c)
{x | x < 2}
>>> latex(c)
'\\left\\{x\\; |\\; x < 2 \\right\\}'
```

Reply:
ping @asmeurer , @jksuom

Reply:
This looks ok to me. The combination `\\; |\\;` could be replaced by `\\mid`.

Reply:
OK